File: 1833-find-the-highest-altitude/find-the-highest-altitude.scala
object Solution {
    def largestAltitude(gain: Array[Int]): Int = {
        def iter(gain: Array[Int], maxAltitude: Int, currentAltitude: Int): Int =
            gain.headOption match
                case Some(num) =>
                    val updatedAltitude = currentAltitude + num
                    val updatedMaxAltitude = maxAltitude max updatedAltitude
                    iter(gain.tail, updatedMaxAltitude, updatedAltitude)
                case None      => maxAltitude

        iter(gain, 0, 0)
    }
}

File: 1833-find-the-highest-altitude/find-the-highest-altitude.java
public class Solution {
    public int largestAltitude(int[] gain) {
        int currentAltitude = 0;
        int maxAltitude = 0;

        for(int i : gain) {
            currentAltitude += i;
            maxAltitude = Math.max(currentAltitude, maxAltitude);
        }

        return maxAltitude;
    }
}